### PR TITLE
Allow to specify ID when adding to the FAISS vectorstore.

### LINKED ISSUE #2699: How to delete or update a document within a FAISS index?
Hi,

I have a usecase where i have to fetch Edited posts weekly from community and update the docs within FAISS index. 
is that possible? or do i have to keep deleting and create new index everytime?

Also i use RecursiveCharacterTextSplitter to split docs.

  ```
  loader = DirectoryLoader('./recent_data')
    raw_documents = loader.load()
    #Splitting documents into chunks
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=1000,
        chunk_overlap=200,
    )
    documents = text_splitter.split_documents(raw_documents)
    print(len(documents))
    # Changing source to point to the original document
    for x in documents:
        print(x.metadata["source"])
    # Creating index and saving it to disk
    print("Creating index")
    db_new = FAISS.from_documents(documents, embeddings )
```

this is output if i use `    print(db_new .docstore._dict)`

`{'2d9b6fbf-a44d-46b5-bcdf-b45cd9438a4c': Document(page_content='<p dir="auto">This is a test topic.</p>', metadata={'source': 'recent/https://community.tpsonline.com/topic/587/ignore-test-topic'}), '706dcaf8-f9d9-45b9-bdf4-8a8ac7618229': Document(page_content='What is an SDD?\n\n<p dir="auto">A software design description (a.k.a. software design document or SDD; just design document; also Software Design Specification) is a representation of a software design that is to be used for recording design information, addressing various design concerns, and communicating that information to the different stakeholders.</p>\n\n<p dir="auto">This SDD template represent design w.r.t various software viewpoints, where each viewpoint will handle specific concerns of Design. This is based on <strong>ISO 42010 standard</strong>.</p>\n\nIntroduction\n\n<p dir="auto">[Name/brief description of feature for which SDD is being Produced]</p>\n\n1. Context Viewpoint\n\n<p dir="auto">[Describes the relationships, dependencies, and interactions between the system and its environment ]</p>\n\n1.1 Use Cases\n\n1.1.1  AS IS (Pre Condition)\n\n1.1.2 TO - BE (Post Condition)\n\n1.2 System Context  View\n\n1.2.1 - AS IS (Pre Condition)\n\n1.2.2 TO - BE (Post Condition)\n\n2. Logical Viewpoint', metadata={'source': 'recent/https://community.tpsonline.com/topic/586/software-design-description-sdd-template'}), '4d6d4e6b-01ee-46bb-ae06-84514a51baf2': Document(page_content='1.1 Use Cases\n\n1.1.1  AS IS (Pre Condition)\n\n1.1.2 TO - BE (Post Condition)\n\n1.2 System Context  View\n\n1.2.1 - AS IS (Pre Condition)\n\n1.2.2 TO - BE (Post Condition)\n\n2. Logical Viewpoint\n\n<p dir="auto">[The purpose of the Logical viewpoint is to elaborate existing and designed types and their implementations as classes and interfaces with their structural static relationships]</p>\n\n2.1 Class Diagram\n\n2.1.1 AS - IS (Pre Condition)\n\n2.1.2 TO - BE (Post Condition)\n\n2.1.2.1 Class Interfaces and description\n\n<p dir="auto">[Below is being presented as an example]<br />\n\n[This section should tell about the responsibility of each class method and their parameters too if required]</p>\n\n2.1.2.1.1 IRenewProcess\n\nMethod\n\nDescription\n\nprocessRenewal\n\nMethod to process renewal of a given cardEntity. Each concrete class that will implement the interface will implement its own version of renewal steps\n\n2.1.2.1.1 RenewStrategyContext (static class)\n\nMethod\n\nDescription\n\n(private)getRenewalMethod', metadata={'source': 'recent/https://community.tpsonline.com/topic/586/software-design-description-sdd-template'})}`

so will i be able to update docs within index or is it just not possible?

### LINKED ISSUE #5065: FAISS should allow you to specify id when using add_text
### System Info

langchain 0.0.173
faiss-cpu 1.7.4

python 3.10.11
Void linux

### Who can help?

@hwchase17 

### Information

- [ ] The official example notebooks/scripts
- [X] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [X] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

It's a logic error in langchain.vectorstores.faiss.__add()

https://github.com/hwchase17/langchain/blob/0c3de0a0b32fadb8caf3e6d803287229409f9da9/langchain/vectorstores/faiss.py#L94-L100
https://github.com/hwchase17/langchain/blob/0c3de0a0b32fadb8caf3e6d803287229409f9da9/langchain/vectorstores/faiss.py#L118-L126

The id is not possible to specify as a function argument. This makes it impossible to detect duplicate additions, for instance.

### Expected behavior

It should be possible to specify id of inserted documents / texts using the add_documents / add_texts methods, as it is in the Chroma object's methods.

As a side-effect this ability would also fix the inability to remove duplicates (see https://github.com/hwchase17/langchain/issues/2699 and https://github.com/hwchase17/langchain/issues/3896 ) by the approach of using ids unique to the content (I use a hash, for example).

### PR DESCRIPTION
**Pull request by atisharma**
# Allow to specify ID when adding to the FAISS vectorstore

This change allows unique IDs to be specified when adding documents / embeddings to a faiss vectorstore.

- This reflects the current approach with the chroma vectorstore.
- It allows rejection of inserts on duplicate IDs
- will allow deletion / update by searching on deterministic ID (such as a hash).
- If not specified, a random UUID is generated (as per previous behaviour, so non-breaking).

This commit fixes #5065 and #3896 and should fix #2699 indirectly. I've tested adding and merging.

Kindly tagging @Xmaster6y @dev2049 for review.

### CODE REVIEW

**Comment by Xmaster6y** [comment]
I think that all the `add` should have the extra id parameter.

Specifying the ids in the `from` class methods should also be possible.

**Comment by atisharma** [comment]
Why?

They all pass through the `ids` argument to the underlying `__add` or `__from`.

Is it for the documentation?

**Comment by atisharma** [comment]
I added the change you suggested.

**Comment by Xmaster6y** [comment]
`__add` is not meant to be used outside the class as a "private method"

File: langchain/vectorstores/faiss.py [hunk 1]
```diff
@@ -129,13 +132,15 @@ def add_texts(
...
         metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
         **kwargs: Any,
     ) -> List[str]:
         """Run more texts through the embeddings and add to the vectorstore.
 
         Args:
             texts: Iterable of strings to add to the vectorstore.
             metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of unique IDs.
```
Comment (hwchase17) [inline langchain/vectorstores/faiss.py:143]: this should be passed to __add right?

Comment (atisharma) [reply langchain/vectorstores/faiss.py:143]: It is, through the **kwargs.

File: langchain/vectorstores/faiss.py [hunk 2]
```diff
@@ -432,6 +443,7 @@ def from_embeddings(
         text_embeddings: List[Tuple[str, List[float]]],
         embedding: Embeddings,
         metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
```
[the remark this answers is not in the document]
Comment (atisharma) [reply langchain/vectorstores/faiss.py:447]: done

File: langchain/vectorstores/faiss.py [hunk 3]
```diff
@@ -161,6 +167,7 @@ def add_embeddings(
             text_embeddings: Iterable pairs of string and embedding to
                 add to the vectorstore.
             metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of unique IDs.
```
[the remark this answers is not in the document]
Comment (atisharma) [reply langchain/vectorstores/faiss.py:170]: done

File: langchain/vectorstores/faiss.py (hunk 1 above)
Comment (atisharma) [reply langchain/vectorstores/faiss.py:143]: done

**Comment by atisharma** [comment]
> `__add` is not meant to be used outside the class as a "private method".

I know.

> I mean how do you specify the ids if you use add_texts?

This wasn't necessary when the ids were not explicitly listed as arguments for the public access methods, because the ids arguments would pass through to the private `__add` and `__from` methods through the `**kwargs`.

With the change you requested, yes, I should have passed them explicitly.

**Review by hwchase17 — APPROVED** [review]
thanks!

**Pull request merged**